### PR TITLE
Adding a design proposal template and README.

### PR DESCRIPTION
**Pull request by tejal29**
This PR is  to add design proposal template. 

This template can help the core team to get a better understanding of the new feature or any big changes.
Also can serve as a guideline to follow when reviewing big changes.

### CODE REVIEW

File: docs/design_proposals/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+Having a proposal, will likely reduce the back and forth between the contributor
+and the core team. It also makes sure, each new feature or a big change has a
+design review.
+
+For any new feature, config or big changes, please add a design proposal document
+as described in [Design Proposal Template](./design-proposal-template.md).
+
+Once you create a PR with the proposal, someone from the core team will be
+assigned as a design sheppard. The role of the design sheppard will be to make
```
Comment (corneliusweig) [inline docs/design_proposals/README.md]: `sheppard` -> `shepherd`

File: docs/design_proposals/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+sure,
+
+1. The feature/change is within Skaffold Philosophy and not a one off
+   solution for a specific use case.
+2. The feature/change scope is well defined.
+3. When changing any existing feature, the implementation plan adheres to
+   [skaffold deprecation policy](./../../deprecation-policy.md)
+
+Once the proposals in a reasonale shape, we can discuss it in Skaffold bi-weekly
```
Comment (corneliusweig) [inline docs/design_proposals/README.md]:
```suggestion
Once the proposal is in a reasonale shape, we can discuss it in Skaffold bi-weekly
```

File: docs/design_proposals/design-proposal-template.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+## Background
+
+In this section, please mention and describe the new feature, re-design
+or re-factor.
+
+Please provide an rationale covering following points:
+
+1. Why is this required?
+2. If its re-design, What are cons with current implementation?
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]:
```suggestion
2. If it is a re-design, what are the cons with the current implementation?
```

File: docs/design_proposals/design-proposal-template.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+3. Are there any another work-around and if yes, why not keep using it.
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]:
```suggestion
3. Are there any another work-arounds and if yes, why not keep using it.
```

File: docs/design_proposals/design-proposal-template.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+Please describe your solution. Please list any:
+
+* new config changes
+* interface changes
+* design assumptions
+
+For a new config change, please mention:
+
+* If its a backward compatible config change
+* If its a backward compatible config change, is there a migration path?
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]: I do not fully understand what is meant by the migration path. I presume, you do not mean the automatic schema upgrade? Can you elaborate on this one a bit more?

File: docs/design_proposals/design-proposal-template.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+**\<Question\>**
+
+Resolution: Please list the resolution if resolved during the design process or
+specify __Not Yet Resolved__
+
+## Implementation plan
+We have identified, huge PRs go unnoticed for a long time. Small incremental
+changes get reviewed faster and also easier for reviewers.
+
+For a design feature, list a summary of tasks breakdown for e.g.:
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]: I don't think that the differentiation between `design feature` and `re-factor proposal` is relevant here. The bottom line is to provide a summary of smaller tasks which should be submitted as separate pull requests, correct? This applies to both cases, such that making a distinction will only confuse people.

File: docs/design_proposals/design-proposal-template.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+For the example desing proposal to infer artifact sync, some of the smaller task
+could be:
+___
+
+1. Add new config key `infer` to `artifact.sync` and test schema validation.
+2. Add inference logic for docker and examples.
+3. Support both `infer` and user defined map with precedence rules implemented.
+4. Finally, support builder plugins to add sync patterns.
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]: Please also include along the lines
```
5. Adapt user documentation.
```
Because it is easily forgotten :)

File: docs/design_proposals/design-proposal-template.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
+# Title
+
+* Author(s): \<your name\>
+* Design Shepherd: As mentioned in the document here.
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]: Maybe make this more self-explanatory?
```suggestion
* Design Shepherd: Please leave this empty, it will be assigned to a core team member.
```

File: docs/design_proposals/design-proposal-template.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+In this section, please mention and describe the new feature, re-design
+or re-factor.
+
+Please provide an rationale covering following points:
+
+1. Why is this required?
+2. If its re-design, What are cons with current implementation?
+3. Are there any another work-around and if yes, why not keep using it.
+
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]: Maybe add an additional item
```
4. Mention related issues, if there are any.
```

File: docs/design_proposals/design-proposal-template.md (hunk 5 above)
Comment (tejal29) [reply docs/design_proposals/design-proposal-template.md]: thanks. i changed to mention how you are going to deprecate it with links to deprecation policy docs.

File: docs/design_proposals/design-proposal-template.md (hunk 6 above)
Comment (tejal29) [reply docs/design_proposals/design-proposal-template.md]: okies i can remove this. :)

File: docs/design_proposals/design-proposal-template.md [hunk 10]
```diff
@@ -13,7 +17,8 @@ Please provide an rationale covering following points:
 
 1. Why is this required?
 2. If its re-design, What are cons with current implementation?
-3. Are there any another work-around and if yes, why not keep using it.
+3. Is there any another work-around and if yes, why not keep using it.
+4. Mention related issues, if there are any.****
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]: If the `****` is intentional, I don't get what it means

Comment (tejal29) [reply docs/design_proposals/design-proposal-template.md]: no its unintentional.

File: docs/design_proposals/README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
+# Design Proposals process
```
Comment (nkubala) [inline docs/design_proposals/README.md]: `Design Proposal Process`

File: docs/design_proposals/README.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+Hello Contributors!
+
+This document describes the process for proposing a new feature or making any
+big code changes to `skaffold`.
+
+Having a proposal, will likely reduce the back and forth between the contributor
```
Comment (nkubala) [inline docs/design_proposals/README.md]: s/`Having a proposal,`/`Submitting a proposal before a pull request`

File: docs/design_proposals/README.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+and the core team. It also makes sure, each new feature or a big change has a
```
Comment (nkubala) [inline docs/design_proposals/README.md]: s/`It also makes sure,`/`A proposal also ensures`

File: docs/design_proposals/README.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+design review.
+
+For any new feature, config or big changes, please add a design proposal document
+as described in [Design Proposal Template](./design-proposal-template.md).
+
+Once you create a PR with the proposal, someone from the core team will be
+assigned as a design shepherd. The role of the design shepherd will be to make
+sure,
```
Comment (nkubala) [inline docs/design_proposals/README.md]: nit: s/`,`/`:`

File: docs/design_proposals/README.md [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+1. The feature/change is within Skaffold Philosophy and not a one off
```
Comment (nkubala) [inline docs/design_proposals/README.md]: "Skaffold Philosophy" isn't really documented anywhere. I'd probably say something more like `The feature or change is in line with the Skaffold roadmap and the team's general philosophy for the tool`

File: docs/design_proposals/README.md [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+   solution for a specific use case.
+2. The feature/change scope is well defined.
+3. When changing any existing feature, the implementation plan adheres to
+   [skaffold deprecation policy](./../../deprecation-policy.md)
+
+Once the proposal is in a reasonale shape, we can discuss it in Skaffold bi-weekly
```
Comment (nkubala) [inline docs/design_proposals/README.md]: s/`reasonale`/`reasonable`

Comment (nkubala) [inline docs/design_proposals/README.md]: can we rephrase this a little bit? "accept or punt" is a little aggressive :)

`Once the proposal has been approved, we can move discussions to our bi-weekly meetings to address any open concerns, and to reach a final decision on whether or not to accept the feature or change.`

File: docs/design_proposals/design-proposal-template.md [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+* Design Shepherd: \<skaffold-core-team-member\>
+
+    If you are already working with someone mention their name.
+    If not, please leave this empty, it will be assigned to a core team member.
+* Date: \<date\>
+* Status: [Draft/Reviewed/Complete]
+
+## Background
+
+In this section, please mention and describe the new feature, re-design
```
Comment (nkubala) [inline docs/design_proposals/design-proposal-template.md]: s/`re-design`/`redesign`

s/`re-factor`/`refactor`

File: docs/design_proposals/design-proposal-template.md [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+or re-factor.
+
+Please provide an rationale covering following points:
```
Comment (nkubala) [inline docs/design_proposals/design-proposal-template.md]: `Please provide a brief explanation for the following questions: `

File: docs/design_proposals/design-proposal-template.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+1. Why is this required?
+2. If its re-design, What are cons with current implementation?
```
Comment (nkubala) [inline docs/design_proposals/design-proposal-template.md]: `If this is a redesign, what are the drawbacks of the current implementation?`

File: docs/design_proposals/design-proposal-template.md [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+3. Is there any another work-around and if yes, why not keep using it.
```
Comment (nkubala) [inline docs/design_proposals/design-proposal-template.md]: `Is there a known workaround, and if so, what are its drawbacks?`

File: docs/design_proposals/design-proposal-template.md [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+
+Please describe your solution. Please list any:
+
+* new config changes
+* interface changes
+* design assumptions
+
+For a new config change, please mention:
+
+* If its a backward compatible config change ?
```
Comment (nkubala) [inline docs/design_proposals/design-proposal-template.md]: let's move this bullet point into the previous question:

```
If there is a new config change, is it backwards compatible? If not, what is the deprecation policy?
Refer to the [deprecation policy requirements] ./../../deprecation-policy.md#how-do-we-deprecate-things) for details.
```

File: docs/design_proposals/design-proposal-template.md [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+
+Please list any open questions here in the format.
+
+**\<Question\>**
+
+Resolution: Please list the resolution if resolved during the design process or
+specify __Not Yet Resolved__
+
+## Implementation plan
+We have identified, huge PRs go unnoticed for a long time. Small incremental
```
Comment (nkubala) [inline docs/design_proposals/design-proposal-template.md]: `As a team, we've noticed that larger PRs can go unreviewed for long periods of time.`

File: docs/design_proposals/design-proposal-template.md [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+changes get reviewed faster and also easier for reviewers.
+
+For a design feature, list a summary of tasks breakdown for e.g.:
+For the example desing proposal to infer artifact sync, some of the smaller task
```
Comment (nkubala) [inline docs/design_proposals/design-proposal-template.md]: s/`desing`/`design`

`For the example artifact sync proposal, some of the smaller tasks could be:`

File: docs/design_proposals/design-proposal-template.md [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+2. Add inference logic for docker and examples.
+3. Support both `infer` and user defined map with precedence rules implemented.
+4. Finally, support builder plugins to add sync patterns.
+
+___
+
+
+## Integration test plan
+
+Please describe what new test cases are you going to consider.
```
Comment (nkubala) [inline docs/design_proposals/design-proposal-template.md]: s/`are you`/`you are`

File: docs/design_proposals/README.md (hunk 16 above)
Comment (tejal29) [reply docs/design_proposals/README.md]: thanks! sounds much better now!

File: docs/design_proposals/README.md [hunk 25]
```diff
@@ -0,0 +1,27 @@
+# Design Proposal Process
+
+Hello Contributors!
+
+This document describes the process for proposing a new feature or making any
+big code changes to `skaffold`.
+
+Submitting a proposal before a pull request, will likely reduce the back and
```
Comment (corneliusweig) [inline docs/design_proposals/README.md]:
```suggestion
Submitting a proposal before a pull request will likely reduce the back and
```

File: docs/design_proposals/README.md [hunk 26]
```diff
@@ -0,0 +1,27 @@
...
+
+Once you create a PR with the proposal, someone from the core team will be
+assigned as a design shepherd. The role of the design shepherd will be to make
+sure:
+
+1. The feature/change is within the Skaffold roadmap and the team's general
+   philosophy for the tool and not a one off solution for a specific use case.
+2. The feature/change scope is well defined.
+3. When changing any existing feature, the implementation plan adheres to
+   [skaffold deprecation policy](./../../deprecation-policy.md)
```
Comment (corneliusweig) [inline docs/design_proposals/README.md]:
```suggestion
   [skaffold deprecation policy](./../../deprecation-policy.md).
```

File: docs/design_proposals/README.md [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+
+Once the proposal has been approved, we can move discussions to our bi-weekly
+meetings to address any open concerns,and to reach a final decision on whether
```
Comment (corneliusweig) [inline docs/design_proposals/README.md]:
```suggestion
meetings to address any open concerns, and to reach a final decision on whether
```

File: docs/design_proposals/design-proposal-template.md [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+
+```yaml
+sync:
+  '*.js': app/
+```
+
+This is error prone and unnecessarily hard to use, because the destination is
+already contained in the Dockerfile for docker build. (see #1166, #1581).
+In addition, the syncing needs to handle special cases for globbing and often
+requires a long list of sync patterns (#1807)
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]:
```suggestion
requires a long list of sync patterns (#1807).
```

File: docs/design_proposals/design-proposal-template.md [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+directly to pods. So far, artifact sync requires a specification of sync
+patterns like
+
+```yaml
+sync:
+  '*.js': app/
+```
+
+This is error prone and unnecessarily hard to use, because the destination is
+already contained in the Dockerfile for docker build. (see #1166, #1581).
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]: no need to reproduce my typos :)
```suggestion
already contained in the Dockerfile for docker build (see #1166, #1581).
```

File: docs/design_proposals/design-proposal-template.md [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+
+* new config changes
+* interface changes
+* design assumptions
+
+For a new config change, please mention:
+
+* Is it backwards compatible? If not, what is the deprecation policy?
+  Refer to the [deprecation policy requirements.](./../../deprecation-policy.md#how-do-we-deprecate-things)
+  for details)
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]:
```suggestion
  for details.
```

File: docs/design_proposals/design-proposal-template.md [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+Please list any open questions here in the format.
+
+**\<Question\>**
+
+Resolution: Please list the resolution if resolved during the design process or
+specify __Not Yet Resolved__
+
+## Implementation plan
+As a team, we've noticed that larger PRs can go unreviewed for long periods of
+time. Small incremental changes get reviewed faster and also easier for reviewers.
```
Comment (corneliusweig) [inline docs/design_proposals/design-proposal-template.md]:
```suggestion
time. Small incremental changes get reviewed faster and are also easier for reviewers.
```

File: docs/design_proposals/README.md [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+Hello Contributors!
+
+This document describes the process for proposing a new feature or making any
+big code changes to `skaffold`.
+
+Submitting a proposal before a pull request will likely reduce the back and
+forth between the contributor and the core team. A proposal also ensures, each
+new feature or a big change has a design review.
+
+For any new feature, config or big changes, please add a design proposal document
```
Comment (balopat) [inline docs/design_proposals/README.md]:
```suggestion
For any new feature, config or large change, please add a design proposal document
```

File: docs/design_proposals/README.md [hunk 33]
```diff
@@ -0,0 +1,27 @@
...
+as described in [Design Proposal Template](./design-proposal-template.md).
+
+Once you create a PR with the proposal, someone from the core team will be
```
Comment (balopat) [inline docs/design_proposals/README.md]:
```suggestion
Once you create a PR with the proposal, someone from the maintainers will be
```

File: docs/design_proposals/README.md [hunk 34]
```diff
@@ -0,0 +1,27 @@
...
+assigned as a design shepherd. The role of the design shepherd will be to make
+sure:
+
+1. The feature/change is within the Skaffold roadmap and the team's general
```
Comment (balopat) [inline docs/design_proposals/README.md]:
```suggestion
1. The feature/change is aligned with the Skaffold roadmap and the team's general
```

File: docs/design_proposals/design-proposal-template.md [hunk 35]
```diff
@@ -0,0 +1,83 @@
...
+## Background
+
+In this section, please mention and describe the new feature, redesign
+or refactor.
+
+Please provide a brief explanation for the following questions:
+
+1. Why is this required?
+2. If this is a redesign, what are the drawbacks of the current implementation?
+3. Is there any another work-around, and if so, what are its drawbacks?
```
Comment (balopat) [inline docs/design_proposals/design-proposal-template.md]:
```suggestion
3. Is there any another workaround, and if so, what are its drawbacks?
```

File: docs/design_proposals/README.md [hunk 36]
```diff
@@ -0,0 +1,27 @@
+# Design Proposal Process
+
+Hello Contributors!
+
+This document describes the process for proposing a new feature or making any
+big code changes to `skaffold`.
```
[the remark this answers is not in the document]
Comment (tejal29) [reply docs/design_proposals/README.md]: will change this to "large" code changes. "larger implies large than X "

File: docs/design_proposals/design-proposal-template.md [hunk 37]
```diff
@@ -0,0 +1,83 @@
+# Title
+
+* Author(s): \<your name\>
+* Design Shepherd: \<skaffold-core-team-member\>
+
+    If you are already working with someone mention their name.
+    If not, please leave this empty, it will be assigned to a core team member.
+* Date: \<date\>
+* Status: [Draft/Reviewed/Complete]
```
Comment (balopat) [inline docs/design_proposals/design-proposal-template.md]: I think Status should be Reviewed / Under implementation / Complete / Cancelled - as the Draft phase is implicit while it's in the PR phase. We should also explain what these are ,e.g.:

`Reviewed` means that the proposal PR has been merged 
`Under implementation` means that the plans is being executed by actual work
`Complete` means that there is no changes expected in this design anymore, if 
`Cancelled` means that we learned something during or before implementation that caused this proposal to be cancelled. 

I would add a `Links` section as well, to collect relevant PRs and issues too. 

What do you think?

File: docs/design_proposals/README.md [hunk 38]
```diff
@@ -0,0 +1,35 @@
...
+
+Hello Contributors!
+
+This document describes the process for proposing a new feature or making any
+large code changes to `skaffold`. By large we mean,  large in impact or large in
+size.
+
+An example for large impact changes would be:
+
+1. Introduce Templating, or
```
Comment (balopat) [inline docs/design_proposals/README.md]:
```suggestion
1. Introduce templating, or
```

File: docs/design_proposals/README.md [hunk 39]
```diff
@@ -0,0 +1,35 @@
+# Design Proposal Process
+
+Hello Contributors!
+
+This document describes the process for proposing a new feature or making any
+large code changes to `skaffold`. By large we mean,  large in impact or large in
+size.
+
+An example for large impact changes would be:
```
Comment (balopat) [inline docs/design_proposals/README.md]:
```suggestion
A examples for large impact changes would be:
```

File: docs/design_proposals/design-proposal-template.md (hunk 37 above)
Comment (tejal29) [reply docs/design_proposals/design-proposal-template.md]: > I would add a Links section as well, to collect relevant PRs and issues too.

I think we have mentioned that in the `Background` bullet 4.

Comment (tejal29) [reply docs/design_proposals/design-proposal-template.md]: > I think Status should be Reviewed / Under implementation / Complete / Cancelled - as the Draft phase is implicit while it's in the PR phase. We should also explain what these are ,e.g.:

So @balopat along your lines but a little different, 

1. In Review: The proposal PR has not been reviewed yet by the core team.
2. Accepted: The proposal PR has been accepted and ready for implementation.
3. Not Accepted: The proposal PR was not accepted by the core team.
4. Under implementation: An accepted proposal is being implemented by actual work.
5. Cancelled: During or before implementation the proposal was cancelled.
   It could be due to:

   * other features added which made the current design proposal obselete.
   * no longer a priority.

6. Complete: This feature/change is implemented.
